File: Algorithms/src/leetcode/longestValidParentheses.c
/*
感谢帮忙测试。我的已经通过了

基本思想是用一个stack s0，遇到'('就push当前的index，遇到')'就pop，说明发现一
个'('-')' pair
此时left和right分别记录这个pair的左边'('的位置和右边')'的位置
同时还有一个额外的s1，记录当前已经遇到的所有pair的左右位置

每次发现一个新的pair (lnew, rnew)，就与s1栈顶的pair(假设是lold, rold)进行比较
如果lnew == rold+1，说明他们是()()这种情况，所以需要合并
如果lnew < rold+1，说明是(())这种情况，也需要合并
一旦进行了第一个合并，就需要继续检查s1直到桟为空或者不能继续合并为止，应付()
(())这种情况

下面是我在leetcode上用的code
*/
#include <stack>
using namespace std;
class Solution {
    public:
        int longestValidParentheses(string s) {
            Start typing your C/C++ solution below
                // DO NOT write int main() function
                stack<unsigned> s0, s1;
            for (unsigned i = 0; i < s.size(); i++) {
                if ('(' == s[i]) {
                    s0.push(i);
                } else {
                    if (!s0.empty()) {
                        unsigned left = s0.top(); s0.pop();
                        unsigned right = i;
                        while (!s1.empty()) {
                            if (s1.empty() || (s1.top()+1 < left)) {
                                // nothing to merge
                                break;
                            } else if (s1.top()+1 == left) {
                                s1.pop();
                                left = s1.top(); s1.pop();
                            } else if (s1.top()+1 > left) {
                                s1.pop(); s1.pop();
                            }
                        }
                        s1.push(left);
                        s1.push(right);
                    }
                }
            }
            unsigned max_len = 0, max_cnt = 1;
            while (!s1.empty()) {
                unsigned left, right;
                right = s1.top(); s1.pop();
                left = s1.top(); s1.pop();
                unsigned len = right-left+1;
                if (len > max_len) {
                    max_len = len; max_cnt = 1;
                } else if (len == max_len) {
                    max_cnt++;
                }
            }
            return max_len;
        }
};

File: Algorithms/src/leetcode/longestValidParentheses.java
// Fail test cases
// "()(()" 4 2
// )()())()()( 8 4
//可以用stack 和 dp。不过我也是在有提示的情况下做出来的。我本来是hbase的思路。
//但是做不对。没想到你那个思路。你这么快能想到正确的解法还是挺赞的。

/*
Given a string containing just the characters '(' and ')', find the length of the longest valid (well-formed) parentheses substring.

For "(()", the longest valid parentheses substring is "()", which has length = 2.

Another example is ")()())", where the longest valid parentheses substring is "()()", which has length = 4.
*/
public class test {
    public static void main(String[] args) {
        Scanner in=new Scanner(System.in);
        String str=in.next();

        int[] dp=new int[str.length()];
        Arrays.fill(dp,-1);
        Stack<Integer> stack=new Stack<Integer>();
        int max=0;
        int maxcount=0;

        for(int i=0;i<str.length();i++) {
            char c=str.charAt(i);
            if(c=='(') {
                stack.add(i);
            } else {
                if(!stack.isEmpty()) {
                    int m=stack.pop();
                    if(m>0 && dp[m-1]!=-1)
                        dp[i]=Math.min(dp[m-1],m);
                    else
                        dp[i]=m;

                    int len=i-dp[i]+1;
                    if(len>max) {
                        max=len;
                        maxcount=1;
                    } else if(len==max)
                        maxcount++;
                }
            }
        }

        if(max>0)
            System.out.println(max+" "+maxcount);
        else
            System.out.println("0 1");
    }
}

/*
import java.util.*;
public class Solution {
    public int longestValidParentheses(String s) {
        // Start typing your Java solution below
        // DO NOT write main() function
        Stack<Character> st = new Stack<Character>();
        int no = 0;
        for (int i=0; i<s.length(); i++) {
            char c = s.charAt(i);
            if (c == '(')
                st.push(c);
            
            if (c == ')' ) {
                if (!st.empty() && st.peek() == '(') {
                    no +=2 ;
                    st.pop();
                }
            }
        }
        return no;
    }
}
*/
